portfolio/src/main/webapp/script.js

```diff
@@ -12,17 +12,220 @@
+function mergeLists() {
+  let out = [];
+
+  Object.keys(FACTS).forEach(key => {
+    out = out.concat(FACTS[key]);
+  });
+  return out;
+}
+
+/* Finds all occurances of substring in the values of arr and returns a list containing
+ * the full strings in arr. Not that this ignores case for searching purposes */
+function findSubstrings(arr, substring) {
+  let out = [];
+
+  substring = substring.toLowerCase();
+  arr.forEach(value => {
+    if (value.toLowerCase().includes(substring)) {
+      out.push(value);
+    }
+  });
+
+  return out;
+}
+
+// Appends a new HTML element to the given parent ID with the given information
+function appendElement(parentId, tagName, innerHtml, elementId = undefined, onclick = undefined) {
+  let tag = document.createElement(tagName);
+  let parent = document.getElementById(parentId);
+
+  tag.innerHTML = innerHtml;
+  if (elementId !== undefined) {
+    tag.id = elementId;
+  }
+  if(onclick !== undefined) {
+    tag.onclick = onclick;
+  }
+  parent.appendChild(tag);
+}
+
+// Removes the given element from the DOM
+function removeElement(elementId) {
+  document.getElementById(elementId).remove();
+}
+
+// Removes all the children from the parent element
+function deleteChildren(parentId) {
+  document.getElementById(parentId).innerHTML = '';
+}
+
+// Retrieves the specified property from the given html element id
+function retrieveProperty(elementId, propertyName) {
+  return document.getElementById(elementId)[propertyName];
+}
+
+/* Put a random fact from the given list into the appropriate container */
+function generateFact() {
+  let selection = retrieveProperty(FAVORITE_CATEGORY_ID, TEXT_SELECTION);
+  let options;
+
+  /* Check if we need to merge all lists */
+  if (selection === MERGE_ALL_LISTS) {
+    options = FACTS_NAMES;
+  } else {
+    options = FACTS[selection];
+  }
+
+  let fact = randomElement(options);
+  //replace the old fact with the one we just generated
+  deleteChildren(RANDOM_FACT_OUTPUT_ID);
+  appendElement(RANDOM_FACT_OUTPUT_ID, 'p', fact);
+}
+
+/* Checks user input against all the facts above and shows options similar
+ * what they typed. */
+function checkAgree() {
+  let input = retrieveProperty(USER_FAVORITE_INPUT_ID, TEXT_SELECTION);
+
+  //remove any current contents so we can insert the results of this search
+  deleteChildren(FAVORITE_SEARCH_RESULT_ID);
+
+  //first check if input exactly matches a fact, ignoring case
+  if (FACTS_NAMES_LOWERCASE.includes(input.toLowerCase())) {
+    appendElement(FAVORITE_SEARCH_RESULT_ID, 'p', FAVORITE_EXACT_MATCH);
+  } else {
+    //check if input is a substring of all the facts
+    let substrings = findSubstrings(FACTS_NAMES, input);
+
+    //Show the user each possibilty
+    if (substrings.length > 0) {
+      //show the tip so users know what to do
+      appendElement(FAVORITE_SEARCH_RESULT_ID, 'p', FAVORITE_SEARCH_TIP);
+
+      //create a list to hold the results and populate it
```

Comment:
"// Create a list..." is the correct format, please make sure all the comments in this file follow this pattern.